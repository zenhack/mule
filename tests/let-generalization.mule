#XFAIL: type formatting
let f = fn x. x in
  { a = f (fn y. y)
  , b = f (Ok {})
  }
