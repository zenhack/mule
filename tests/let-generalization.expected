
Desugared:

	(letrec (types ()) (values ((f (fn x x))))
	 (letrec (types ()) (values ((a (f (fn y y))) (b (f ((fn x (Ok x)) {})))))
	  (.= a a (.= b b {}))))

inferred type:

	(record (types ())
	 (values
	  ((a (all a (a -> a))) (b (union ((Ok (record (types ()) (values ())))))))))

Runtime term:

	(LetRec
	 (((0 (Lam (0 () (Var 0)))))
	  (LetRec
	   (((3 (App ((Var 2) (Lam (0 () (Var 0))))))
	     (3 (App ((Var 2) (App ((Lam (0 () (Ctor (Ok (Var 0))))) (Record ())))))))
	    (Update (old (Update (old (Record ())) (label b) (field (Var 1))))
	     (label a) (field (Var 0)))))))

Evaluated:

	(Record ((a (Lam (0 () (Var 0)))) (b (Ctor (Ok (Record ()))))))
